### PR TITLE
test(boavizt): add tests for client and api

### PR DESCRIPTION
**Pull request by nikhiln64**
# In this PR
- Added tests around BoaviztAPI module addressing [issue#16](https://github.com/DigitalPebble/spruce/issues/16)
- BoaviztAPIClientTest covers BoaviztAPIClient
- BoaviztAPITest covers BoaviztAPI

```
[INFO] Tests run: 22, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 110.3 s -- in com.digitalpebble.spruce.modules.boavizta.BoaviztAPIClientTest
[INFO] Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 65.08 s -- in com.digitalpebble.spruce.modules.boavizta.BoaviztAPITest

```

Screenshot 2025-08-20 at 1 47 51 PM

### CODE REVIEW

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    void testConstructor() {
+        BoaviztAPIClient testClient = new BoaviztAPIClient("http://custom-host:8080");
+        assertNotNull(testClient);
+    }
+
+    @Test
+    void testConstructorWithNullHost() {
+        // The constructor doesn't actually throw NullPointerException for null host
+        // It just stores the null value, which will cause issues later when making HTTP calls
+        BoaviztAPIClient testClient = new BoaviztAPIClient(null);
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:33]: it would be worth changing the behaviour of the constructor so that an error is thrown

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        assertNotNull(testClient);
+    }
+
+    @Test
+    void testConstructorWithEmptyHost() {
+        BoaviztAPIClient testClient = new BoaviztAPIClient("");
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:39]: same here - it would be worth changing the behaviour of the constructor so that an error is thrown

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    }
+
+    @Test
+    void testConstructorWithHttpsHost() {
+        BoaviztAPIClient testClient = new BoaviztAPIClient("https://secure-host:8443");
+        assertNotNull(testClient);
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithValidInstanceType() throws IOException {
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:50]: given that this test is subsumed by the one below, what is the value of keeping it?

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        }
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithNullInstanceType() throws IOException {
+        // The method doesn't throw NullPointerException for null instance type
+        // It just passes "null" as a string to the API
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, null);
+            assertNotNull(result);
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:95]: Worth fixing this behaviour as part of this PR?

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+                      e.getMessage().contains("Connection refused") ||
+                      e.getMessage().contains("Failed to connect"),
+                      "Expected connection-related error, got: " + e.getMessage());
+        }
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithEmptyInstanceType() throws IOException {
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, "");
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:109]: Worth fixing this behaviour as part of this PR?

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                      e.getMessage().contains("Connection refused") ||
+                      e.getMessage().contains("Failed to connect"),
+                      "Expected connection-related error, got: " + e.getMessage());
+        }
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithWhitespaceInstanceType() throws IOException {
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, "   ");
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:124]: Worth fixing this behaviour as part of this PR?

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+                          e.getMessage().contains("Connection refused") ||
+                          e.getMessage().contains("Failed to connect"),
+                          "Expected connection-related error for " + instanceType + ", got: " + e.getMessage());
+            }
+        }
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithVeryLongInstanceType() throws IOException {
+        String longInstanceType = "t3.micro".repeat(100); // Create a very long instance type
```
[the remark this answers is not in the document]
Comment (jnioche) [reply src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:157]: Tried a review by Copilot to see what is was like. This suggestion can probably be ignored

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+            }
+        }
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithNullProvider() throws IOException {
+        // The method doesn't throw NullPointerException for null provider
+        // It just passes null to the API call
+        try {
+            double[] result = client.getEnergyEstimates(null, "t3.micro");
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:250]: Worth fixing this behaviour as part of the PR?

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPITest.java [hunk 9]
```diff
@@ -2,26 +2,267 @@
...
         Row row = new GenericRowWithSchema(values, schema);
         Row enriched = api.process(row);
-        // missing values comes back as it was
+        
+        // Should return the original row unchanged
+        assertEquals(row, enriched);
+    }
+
+    @Test
+    void testProcessWithNullOperation() {
```
Comment (jnioche) [inline src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPITest.java:77]: given that these tests have a similar behaviour and check that the inputs are not changed, they differ only in their inputs. What about having a parameterized test instead?
https://www.baeldung.com/parameterized-tests-junit-5#6-method

**Review by jnioche — COMMENTED** [review]
thanks for the PR @nikhiln64 
you tests expose some valid improvements that could be done in the handling of constructors. Would be good to implement these as part of the PR

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java (hunk 2 above)
Comment (nikhiln64) [reply src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:39]: Sure.

**Review by jnioche — CHANGES_REQUESTED** [review]
Thanks @nikhiln64 this is a great addition to the project and a better version of the PR.
I have left a few comments and suggestions.
Thanks again for your contribution

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 10]
```diff
@@ -3,71 +3,96 @@
...
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
 
 import java.io.IOException;
+import java.util.stream.Stream;
 
 import static org.junit.jupiter.api.Assertions.*;
 
-public class BoaviztAPIClientTest extends AbstractBoaviztaTest {
```
[the remark this answers is not in the document]
Comment (nikhiln64) [reply src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java]: Real API responses can vary over time, making it difficult to control the returned data. This limits the ability to reliably test error scenarios (e.g., 403 errors, timeouts) and prevents covering edge cases that the real API may not support.

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class BoaviztAPIClientTest {
+
+    private MockWebServer mockWebServer;
+    private BoaviztAPIClient client;
+    private static final String TEST_HOST = "http://localhost:5000";
+
+    @BeforeEach
```
[the remark this answers is not in the document]
Comment (nikhiln64) [reply src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java]: Segregated them. lmk wyt.

File: src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jnioche) [reply src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java]: I get your point but then why not apply the same approach to the other test class BoaviztAPITest?
Not having to have Docker for testing would be a plus

**Comment by jnioche** [comment]
What would you think about changing AbstractBoaviztaTest so that the container is started once and for all and not for every test? This would reduce the time taken by the tests and reduce the risk of getting errors

**Comment by nikhiln64** [comment]
@jnioche Addressed your comments and here are the improvements after refactoring:
Specific results:
* Previous approach: ~34 seconds for BoaviztAPI tests
* Current approach: 0.092 seconds for BoaviztAPI tests

Total results:


Metric | Before Refactoring | After Refactoring | Improvement
-- | -- | -- | --
BoaviztAPI Tests | ~34s | 0.092s | 370x faster
Total Test Suite | ~40s+ | 3.225s | 12x faster




Screenshot 2025-08-31 at 8 19 11 AM

**Comment by jnioche** [comment]
thanks @nikhiln64 
Since you have moved the tests to use the mocks, I don't think we need to keep _AbstractBoavizta*_ classes at all, the dependency on org.testcontainers can also be removed. What do you think?

**Comment by jnioche** [comment]
> thanks @nikhiln64 Since you have moved the tests to use the mocks, I don't think we need to keep _AbstractBoavizta*_ classes at all, the dependency on org.testcontainers can also be removed. What do you think?

I will do that separately.

Thanks for this PR @nikhiln64 and your patience with my comments and suggestions, much appreciated!

Have you used Spruce at all?
